// @ethereumjs/blockchain@7.3.0/dist/esm/consensus/ethash.js
import { ConsensusAlgorithm } from '@ethereumjs/common';
import { Ethash } from '@ethereumjs/ethash';
/**
 * This class encapsulates Ethash-related consensus functionality when used with the Blockchain class.
 */
export class EthashConsensus {
    constructor() {
        this.algorithm = ConsensusAlgorithm.Ethash;
    }
    async validateConsensus(block) {
        if (!this._ethash) {
            throw new Error('blockchain not provided');
        }
        const valid = await this._ethash.verifyPOW(block);
        if (!valid) {
            throw new Error('invalid POW');
        }
    }
    /**
     * Checks that the block's `difficulty` matches the canonical difficulty of the parent header.
     * @param header - header of block to be checked
     */
    async validateDifficulty(header) {
        if (!this.blockchain) {
            throw new Error('blockchain not provided');
        }
        const parentHeader = await this.blockchain['_getHeader'](header.parentHash);
        if (header.ethashCanonicalDifficulty(parentHeader) !== header.difficulty) {
            throw new Error(`invalid difficulty ${header.errorStr()}`);
        }
    }
    async genesisInit() { }
    async setup({ blockchain }) {
        this.blockchain = blockchain;
        this._ethash = new Ethash(this.blockchain.db);
    }
    async newBlock() { }
}
