// @ethereumjs/evm@3.1.1/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NobleBLS = exports.Message = exports.MCLBLS = exports.getOpcodesForHF = exports.getActivePrecompiles = exports.EVMErrorMessage = exports.EvmError = exports.EVM = exports.EOF = void 0;
const eof_js_1 = require("./eof.js");
Object.defineProperty(exports, "EOF", { enumerable: true, get: function () { return eof_js_1.EOF; } });
const evm_js_1 = require("./evm.js");
Object.defineProperty(exports, "EVM", { enumerable: true, get: function () { return evm_js_1.EVM; } });
const exceptions_js_1 = require("./exceptions.js");
Object.defineProperty(exports, "EVMErrorMessage", { enumerable: true, get: function () { return exceptions_js_1.ERROR; } });
Object.defineProperty(exports, "EvmError", { enumerable: true, get: function () { return exceptions_js_1.EvmError; } });
const message_js_1 = require("./message.js");
Object.defineProperty(exports, "Message", { enumerable: true, get: function () { return message_js_1.Message; } });
const index_js_1 = require("./opcodes/index.js");
Object.defineProperty(exports, "getOpcodesForHF", { enumerable: true, get: function () { return index_js_1.getOpcodesForHF; } });
const index_js_2 = require("./precompiles/index.js");
Object.defineProperty(exports, "MCLBLS", { enumerable: true, get: function () { return index_js_2.MCLBLS; } });
Object.defineProperty(exports, "NobleBLS", { enumerable: true, get: function () { return index_js_2.NobleBLS; } });
Object.defineProperty(exports, "getActivePrecompiles", { enumerable: true, get: function () { return index_js_2.getActivePrecompiles; } });
__exportStar(require("./logger.js"), exports);
