// @ethereumjs/evm@3.1.1/dist/cjs/precompiles/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ripemdPrecompileAddress = exports.precompiles = exports.precompileEntries = exports.NobleBLS = exports.MCLBLS = exports.getPrecompileName = exports.getActivePrecompiles = void 0;
const common_1 = require("@ethereumjs/common");
const util_1 = require("@ethereumjs/util");
const _01_ecrecover_js_1 = require("./01-ecrecover.js");
const _02_sha256_js_1 = require("./02-sha256.js");
const _03_ripemd160_js_1 = require("./03-ripemd160.js");
const _04_identity_js_1 = require("./04-identity.js");
const _05_modexp_js_1 = require("./05-modexp.js");
const _06_ecadd_js_1 = require("./06-ecadd.js");
const _07_ecmul_js_1 = require("./07-ecmul.js");
const _08_ecpairing_js_1 = require("./08-ecpairing.js");
const _09_blake2f_js_1 = require("./09-blake2f.js");
const _0a_kzg_point_evaluation_js_1 = require("./0a-kzg-point-evaluation.js");
const _0b_bls12_g1add_js_1 = require("./0b-bls12-g1add.js");
const _0c_bls12_g1mul_js_1 = require("./0c-bls12-g1mul.js");
const _0d_bls12_g1msm_js_1 = require("./0d-bls12-g1msm.js");
const _0e_bls12_g2add_js_1 = require("./0e-bls12-g2add.js");
const _0f_bls12_g2mul_js_1 = require("./0f-bls12-g2mul.js");
const _10_bls12_g2msm_js_1 = require("./10-bls12-g2msm.js");
const _11_bls12_pairing_js_1 = require("./11-bls12-pairing.js");
const _12_bls12_map_fp_to_g1_js_1 = require("./12-bls12-map-fp-to-g1.js");
const _13_bls12_map_fp2_to_g2_js_1 = require("./13-bls12-map-fp2-to-g2.js");
const index_js_1 = require("./bls12_381/index.js");
Object.defineProperty(exports, "MCLBLS", { enumerable: true, get: function () { return index_js_1.MCLBLS; } });
Object.defineProperty(exports, "NobleBLS", { enumerable: true, get: function () { return index_js_1.NobleBLS; } });
var PrecompileAvailabilityCheck;
(function (PrecompileAvailabilityCheck) {
    PrecompileAvailabilityCheck[PrecompileAvailabilityCheck["EIP"] = 0] = "EIP";
    PrecompileAvailabilityCheck[PrecompileAvailabilityCheck["Hardfork"] = 1] = "Hardfork";
})(PrecompileAvailabilityCheck || (PrecompileAvailabilityCheck = {}));
const ripemdPrecompileAddress = '0000000000000000000000000000000000000003';
exports.ripemdPrecompileAddress = ripemdPrecompileAddress;
const precompileEntries = [
    {
        address: '0000000000000000000000000000000000000001',
        check: {
            type: PrecompileAvailabilityCheck.Hardfork,
            param: common_1.Hardfork.Chainstart,
        },
        precompile: _01_ecrecover_js_1.precompile01,
        name: 'ECRECOVER (0x01)',
    },
    {
        address: '0000000000000000000000000000000000000002',
        check: {
            type: PrecompileAvailabilityCheck.Hardfork,
            param: common_1.Hardfork.Chainstart,
        },
        precompile: _02_sha256_js_1.precompile02,
        name: 'SHA256 (0x02)',
    },
    {
        address: '0000000000000000000000000000000000000003',
        check: {
            type: PrecompileAvailabilityCheck.Hardfork,
            param: common_1.Hardfork.Chainstart,
        },
        precompile: _03_ripemd160_js_1.precompile03,
        name: 'RIPEMD160 (0x03)',
    },
    {
        address: '0000000000000000000000000000000000000004',
        check: {
            type: PrecompileAvailabilityCheck.Hardfork,
            param: common_1.Hardfork.Chainstart,
        },
        precompile: _04_identity_js_1.precompile04,
        name: 'Identity (0x04)',
    },
    {
        address: '0000000000000000000000000000000000000005',
        check: {
            type: PrecompileAvailabilityCheck.Hardfork,
            param: common_1.Hardfork.Byzantium,
        },
        precompile: _05_modexp_js_1.precompile05,
        name: 'MODEXP (0x05)',
    },
    {
        address: '0000000000000000000000000000000000000006',
        check: {
            type: PrecompileAvailabilityCheck.Hardfork,
            param: common_1.Hardfork.Byzantium,
        },
        precompile: _06_ecadd_js_1.precompile06,
        name: 'ECADD (0x06)',
    },
    {
        address: '0000000000000000000000000000000000000007',
        check: {
            type: PrecompileAvailabilityCheck.Hardfork,
            param: common_1.Hardfork.Byzantium,
        },
        precompile: _07_ecmul_js_1.precompile07,
        name: 'ECMUL (0x07)',
    },
    {
        address: '0000000000000000000000000000000000000008',
        check: {
            type: PrecompileAvailabilityCheck.Hardfork,
            param: common_1.Hardfork.Byzantium,
        },
        precompile: _08_ecpairing_js_1.precompile08,
        name: 'ECPAIR (0x08)',
    },
    {
        address: '0000000000000000000000000000000000000009',
        check: {
            type: PrecompileAvailabilityCheck.Hardfork,
            param: common_1.Hardfork.Istanbul,
        },
        precompile: _09_blake2f_js_1.precompile09,
        name: 'BLAKE2f (0x09)',
    },
    {
        address: '000000000000000000000000000000000000000a',
        check: {
            type: PrecompileAvailabilityCheck.EIP,
            param: 4844,
        },
        precompile: _0a_kzg_point_evaluation_js_1.precompile0a,
        name: 'KZG (0x0a)',
    },
    {
        address: '000000000000000000000000000000000000000b',
        check: {
            type: PrecompileAvailabilityCheck.EIP,
            param: 2537,
        },
        precompile: _0b_bls12_g1add_js_1.precompile0b,
        name: 'BLS12_G1ADD',
    },
    {
        address: '000000000000000000000000000000000000000c',
        check: {
            type: PrecompileAvailabilityCheck.EIP,
            param: 2537,
        },
        precompile: _0c_bls12_g1mul_js_1.precompile0c,
        name: 'BLS12_G1MUL',
    },
    {
        address: '000000000000000000000000000000000000000d',
        check: {
            type: PrecompileAvailabilityCheck.EIP,
            param: 2537,
        },
        precompile: _0d_bls12_g1msm_js_1.precompile0d,
        name: 'BLS12_G1MSM',
    },
    {
        address: '000000000000000000000000000000000000000e',
        check: {
            type: PrecompileAvailabilityCheck.EIP,
            param: 2537,
        },
        precompile: _0e_bls12_g2add_js_1.precompile0e,
        name: 'BLS12_G2ADD',
    },
    {
        address: '000000000000000000000000000000000000000f',
        check: {
            type: PrecompileAvailabilityCheck.EIP,
            param: 2537,
        },
        precompile: _0f_bls12_g2mul_js_1.precompile0f,
        name: 'BLS12_G2MUL',
    },
    {
        address: '0000000000000000000000000000000000000010',
        check: {
            type: PrecompileAvailabilityCheck.EIP,
            param: 2537,
        },
        precompile: _10_bls12_g2msm_js_1.precompile10,
        name: 'BLS12_G2MSM',
    },
    {
        address: '0000000000000000000000000000000000000011',
        check: {
            type: PrecompileAvailabilityCheck.EIP,
            param: 2537,
        },
        precompile: _11_bls12_pairing_js_1.precompile11,
        name: 'BLS12_PAIRING',
    },
    {
        address: '0000000000000000000000000000000000000012',
        check: {
            type: PrecompileAvailabilityCheck.EIP,
            param: 2537,
        },
        precompile: _12_bls12_map_fp_to_g1_js_1.precompile12,
        name: 'BLS12_MAP_FP_TO_G1',
    },
    {
        address: '0000000000000000000000000000000000000013',
        check: {
            type: PrecompileAvailabilityCheck.EIP,
            param: 2537,
        },
        precompile: _13_bls12_map_fp2_to_g2_js_1.precompile13,
        name: 'BLS12_MAP_FP2_TO_G2',
    },
];
exports.precompileEntries = precompileEntries;
const precompiles = {
    '0000000000000000000000000000000000000001': _01_ecrecover_js_1.precompile01,
    '0000000000000000000000000000000000000002': _02_sha256_js_1.precompile02,
    [ripemdPrecompileAddress]: _03_ripemd160_js_1.precompile03,
    '0000000000000000000000000000000000000004': _04_identity_js_1.precompile04,
    '0000000000000000000000000000000000000005': _05_modexp_js_1.precompile05,
    '0000000000000000000000000000000000000006': _06_ecadd_js_1.precompile06,
    '0000000000000000000000000000000000000007': _07_ecmul_js_1.precompile07,
    '0000000000000000000000000000000000000008': _08_ecpairing_js_1.precompile08,
    '0000000000000000000000000000000000000009': _09_blake2f_js_1.precompile09,
    '000000000000000000000000000000000000000a': _0a_kzg_point_evaluation_js_1.precompile0a,
    '000000000000000000000000000000000000000b': _0b_bls12_g1add_js_1.precompile0b,
    '000000000000000000000000000000000000000c': _0c_bls12_g1mul_js_1.precompile0c,
    '000000000000000000000000000000000000000d': _0d_bls12_g1msm_js_1.precompile0d,
    '000000000000000000000000000000000000000e': _0e_bls12_g2add_js_1.precompile0e,
    '000000000000000000000000000000000000000f': _0f_bls12_g2mul_js_1.precompile0f,
    '0000000000000000000000000000000000000010': _10_bls12_g2msm_js_1.precompile10,
    '0000000000000000000000000000000000000011': _11_bls12_pairing_js_1.precompile11,
    '0000000000000000000000000000000000000012': _12_bls12_map_fp_to_g1_js_1.precompile12,
    '0000000000000000000000000000000000000013': _13_bls12_map_fp2_to_g2_js_1.precompile13,
};
exports.precompiles = precompiles;
function getActivePrecompiles(common, customPrecompiles) {
    const precompileMap = new Map();
    if (customPrecompiles) {
        for (const precompile of customPrecompiles) {
            precompileMap.set((0, util_1.bytesToUnprefixedHex)(precompile.address.bytes), 'function' in precompile ? precompile.function : undefined);
        }
    }
    for (const entry of precompileEntries) {
        if (precompileMap.has(entry.address)) {
            continue;
        }
        const type = entry.check.type;
        if ((type === PrecompileAvailabilityCheck.Hardfork && common.gteHardfork(entry.check.param)) ||
            (entry.check.type === PrecompileAvailabilityCheck.EIP &&
                common.isActivatedEIP(entry.check.param))) {
            precompileMap.set(entry.address, entry.precompile);
        }
    }
    return precompileMap;
}
exports.getActivePrecompiles = getActivePrecompiles;
function getPrecompileName(addressUnprefixedStr) {
    for (const entry of precompileEntries) {
        if (entry.address === addressUnprefixedStr) {
            return entry.name;
        }
    }
}
exports.getPrecompileName = getPrecompileName;
